API/NoDocs/Microsoft.Extensions.Diagnostics.Metrics/CounterAttribute.cs

```diff
@@ -0,0 +1,18 @@
+// Assembly 'Microsoft.Extensions.Diagnostics.ExtraAbstractions'
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Diagnostics.Metrics;
+
+[AttributeUsage(AttributeTargets.Method)]
+[Conditional("CODE_GENERATION_ATTRIBUTES")]
+public sealed class CounterAttribute : Attribute
+{
+    public string? Name { get; set; }
+    public string[]? TagNames { get; }
+    public Type? Type { get; }
+    public CounterAttribute(params string[] tagNames);
+    public CounterAttribute(Type type);
+}
```

Comment:
I'm not a fan of the share of CounterAttribute. I could imagine someone entering [Counter("my.cool.counter.name")] and expecting a counter with that name but instead accidentally calling the `CounterAttribute(params string[] tagNames)` ctor and getting a counter with an inferred name and tag named `my.cool.counter.name` instead. Feels like pit of failure.

The most significant piece of information for a counter is its name. It should be in the constructor. Also, it should always be required. Infering it from a return type doesn't feel good. Thought should be put into the name.

Also, the `Type` property name isn't descriptive. `TagType` is still a little confusing because tags can have a type.

```suggestion
public sealed class CounterAttribute : Attribute
{
    public string? Name { get; }
    public string[]? TagNames { get; set; }
    public Type? TagType { get; set; }
    public CounterAttribute(string name);
}
```

Usage:

```cs
[Counter("my.cool.counter.name", TagNames = [ "one", "two", "etc" ])]
````

If someone sets `TagNames` and `TagType` then throw an error.

Reply:
@geeknoid @dariusclay @iliar-turdushev folks, any thoughts here?
Personally I see the point of the suggested change, but @JamesNK please bear in mind that by default the source-gen will take a type's name as an instrument name, that's why its API was made that way.

Reply:
I'm good with the suggestion, but it will be a breaking change for existing customers, and complicate their migration to dotnet/extensions. Is it OK?

The only change I'd make to the suggestion is mark `name` argument as optional:

```csharp
public CounterAttribute(string? name = null);
```

If `name` is `null` then we infer the name of a counter from the return type of a method.

Reply:
This comment could have been addressed a few weeks back, but now it's too late. We can no longer do breaking changes in dotnet/extensions. So let's improve this incrementally when possible, without breaking things.